### PR TITLE
Stabilize wlr_box API

### PR DESCRIPTION
**Pull request by bl4ckb0ne**
Fixes https://github.com/swaywm/wlroots/issues/3006

Breaking changes:

- Removed unused `wlr_box_from_pixman_box32` and `wlr_box_rotated_bounds` functions.
- Unnecessary `wlr_box.h` includes have been replaced with forward delcarations.

### CODE REVIEW

**Comment by emersion** [comment]
It would be nice to properly document this header while we're at it. There are some important details such as whether the box bounds are excluded or included in the box etc.

**Comment by bl4ckb0ne** [comment]
I though the function names were self explaining enough, will make another pass and merge it in the last commit.

File: include/wlr/types/wlr_box.h [hunk 1]
```diff
@@ -1,42 +1,7 @@
...
-bool wlr_box_contains_point(const struct wlr_box *box, double x, double y);
-
-bool wlr_box_empty(const struct wlr_box *box);
-
-/**
- * Transforms a box inside a `width` x `height` box.
- */
-void wlr_box_transform(struct wlr_box *dest, const struct wlr_box *box,
-	enum wl_output_transform transform, int width, int height);
+#warning "wlr_box has been moved ton wlr/util/box.h"
```
Comment (emersion) [inline include/wlr/types/wlr_box.h:4]: Typo: "to"

File: include/wlr/util/box.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * Note that wlroots does not make an ABI compatibility promise - in the future,
+ * the layout and size of structs used by wlroots may change, requiring code
+ * depending on this header to be recompiled (but not edited).
+ *
+ * Breaking changes are announced by email and follow a 1-year deprecation
+ * schedule. Send an email to ~sircmpwn/wlroots-announce+subscribe@lists.sr.ht
+ * to receive these announcements.
+ */
+#ifndef WLR_UTIL_WLR_BOX_H
```
Comment (emersion) [inline include/wlr/util/box.h:16]: This should be named `WLR_UTIL_BOX_H`: we derive the define name from the file path.

**Comment by emersion** [comment]
> It would be nice to properly document this header while we're at it. There are some important details such as whether the box bounds are excluded or included in the box etc.

Bump: can we document this? It's important IMHO to document that the point at (100, 50) is not contained in a (0, 0, 100, 50) box.

File: include/wlr/util/box.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * - New functions
+ * - New struct members
+ * - New enum members
+ *
+ * Note that wlroots does not make an ABI compatibility promise - in the future,
+ * the layout and size of structs used by wlroots may change, requiring code
+ * depending on this header to be recompiled (but not edited).
+ *
+ * Breaking changes are announced by email and follow a 1-year deprecation
+ * schedule. Send an email to ~sircmpwn/wlroots-announce+subscribe@lists.sr.ht
```
Comment (emersion) [inline include/wlr/util/box.h:13]: The ML has been removed

File: include/wlr/util/box.h [hunk 4]
```diff
@@ -1,15 +1,20 @@
...
+ * Breaking changes are announced in the release notes and follow a 1-year
+ * deprecation schedule.
  */
-#ifndef WLR_USE_UNSTABLE
-#error "Add -DWLR_USE_UNSTABLE to enable unstable wlroots features"
-#endif
-
-#ifndef WLR_TYPES_WLR_BOX_H
-#define WLR_TYPES_WLR_BOX_H
+#ifndef WLR__BOX_H
```
Comment (emersion) [inline include/wlr/util/box.h:15]: This guard should be named `WLR_UTIL_BOX_H` based on the file path.

File: include/wlr/util/box.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ * Breaking changes are announced in the release notes and follow a 1-year
+ * deprecation schedule.
+ */
+#ifndef WLR__BOX_H
+#define WLR__BOX_H
+
+#include <stdbool.h>
+#include <wayland-server-protocol.h>
+
+struct wlr_box {
```
Comment (emersion) [inline include/wlr/util/box.h:21]: Suggested doc comment

```c
/**
 * A box describes a rectangle region in a 2D space.
 *
 * The x and y coordinates are inclusive, and the width and height lengths are
 * exclusive. In other words, the box starts from the coordinates (x, y), and
 * goes up to but not including (x + width, y + height).
 */
```

File: include/wlr/util/box.h [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	int x, y;
+	int width, height;
+};
+
+struct wlr_fbox {
```
Comment (emersion) [inline include/wlr/util/box.h:26]: Suggested doc comment

```c
/**
 * A floating-point box.
 *
 * wlr_fbox has the same semantics as wlr_box.
 */
```

File: include/wlr/util/box.h [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+/*
+ * Gives the intersecting box betweem two wlr_box.
+ * Returns an empty wlr_box the provided wlr_box don't intersect.
+ */
+bool wlr_box_intersection(struct wlr_box *dest, const struct wlr_box *box_a,
+	const struct wlr_box *box_b);
+
+/*
+ * Verifies if a point is contained within the bounds of a given wlr_box.
+ * For example, a point at (100, 50) is not contained in the box (0, 0, 100, 50).
```
Comment (emersion) [inline include/wlr/util/box.h:47]: Maybe give more examples?

```c
 * For example:
 *
 * - A point at (10, 10) is contained in the box (10, 0, 50, 50).
 * - A point at (0, -10) is not contained in the box (10, 0, 50, 50).
 * - A point at (60, 50) is not contained in the box (10, 0, 50, 50).
```

File: include/wlr/util/box.h [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+ */
+bool wlr_box_contains_point(const struct wlr_box *box, double x, double y);
+
+bool wlr_box_empty(const struct wlr_box *box);
```
Comment (emersion) [inline include/wlr/util/box.h:51]: Suggested doc comment

```c
/**
 * Checks whether a box is empty.
 *
 * A box is empty if its width or height is zero or negative.
 */
```

File: include/wlr/util/box.h [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+/*
+ * Finds the closet point within the box bounds
+ * Returns NAN if the box is empty
+ */
+void wlr_box_closest_point(const struct wlr_box *box, double x, double y,
+	double *dest_x, double *dest_y);
+
+/*
+ * Gives the intersecting box betweem two wlr_box.
+ * Returns an empty wlr_box the provided wlr_box don't intersect.
```
Comment (emersion) [inline include/wlr/util/box.h:40]: Typo: missing "if"

File: include/wlr/util/box.h [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+	int width, height;
+};
+
+struct wlr_fbox {
+	double x, y;
+	double width, height;
+};
+
+/*
+ * Finds the closet point within the box bounds
```
Comment (emersion) [inline include/wlr/util/box.h:32]: Typo: "closest"

File: include/wlr/util/box.h [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+/*
+ * Verifies if a point is contained within the bounds of a given wlr_box.
+ * For example, a point at (100, 50) is not contained in the box (0, 0, 100, 50).
+ */
+bool wlr_box_contains_point(const struct wlr_box *box, double x, double y);
+
+bool wlr_box_empty(const struct wlr_box *box);
+
+/**
+ * Transforms a box inside a `width` x `height` box.
```
Comment (emersion) [inline include/wlr/util/box.h:54]: Maybe we should make it clearer that the large box has zero x and y?

```c
/**
 * Transforms a box inside a (0, 0, width, height) box.
 */
```

File: include/wlr/util/box.h (hunk 4 above)
Comment (bl4ckb0ne) [reply include/wlr/util/box.h:15]: Woops, messed up my vim replace motion.

**Review by emersion — APPROVED** [review]
LGTM, thanks!

**Pull request merged**